## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('../Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
column_names = df.columns
print(column_names)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [7]:
# Use the describe function to gather some basic statistics
statistics = df.describe()
print(statistics)

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [17]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [24]:
# What three item categories had the most entries?
value_counts = df['category'].value_counts()
print(value_counts)

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64


In [30]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = df["category"].value_counts().idxmax()
filtered_df = df[df["category"] == most_common_category]
most_common_subcategory = filtered_df["subcategory"].value_counts().idxmax()
print(most_common_subcategory)


bathroom supplies


In [36]:
# Which five clients had the most entries in the data?
client_counts = df['client_id'].value_counts()
clients_sorted = client_counts.sort_values(ascending=False)
top_five_clients = clients_sorted.head(5)
print(top_five_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [37]:
# Store the client ids of those top 5 clients in a list.
top_five_client_ids = top_five_clients.index.tolist()
print(top_five_client_ids)

[33615, 66037, 46820, 38378, 24741]


In [45]:
# How many total units (the qty column) did the client with the most entries order order?
client_total_units = df.groupby('client_id')['qty'].sum()
most_common_client_id = client_total_units.idxmax()
total_units_most_most_common_client = client_total_units.loc[most_common_client_id]
print(total_units_most_most_common_client)

4033302


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [48]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["subtotal"] = df["unit_price"] * df["qty"]
print(df)

           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0       

In [51]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping_price(unit_weight):
    if unit_weight >50:
        return 7 * unit_weight
    else:
        return 10 * unit_weight

df["shipping_price"] = df["unit_weight"].apply(calculate_shipping_price)
print(df)

           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0       

In [54]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df["sales_tax"] = (df["subtotal"]+ df["shipping_price"]) * sales_tax_rate
df["total_price"] = df["subtotal"] + df["shipping_price"] +df["sales_tax"]
print(df)


           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0       

In [56]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost"] = df["unit_cost"] * df["qty"] + df["shipping_price"]
print(df)



           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0       

In [60]:
# Create a column for the profit of each line using line cost and line price

df["profit"] = df["total_price"] - df["line_cost"]
print(df)

           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0       

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [72]:
# Check your work using the totals above
order_id_to_print = "2742071"
filtered_df = df[df["order_id"] == order_id_to_print]
print(order_id_to_print)
for index, row in filtered_df.iterrows():
    print("Line {}: Total_Price:".format (index + 1,
        row["total_price"]   
    ))
                                           


2742071


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [85]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
client_total_qty = df.groupby("client_id")["qty"].sum()
sorted_clients = client_total_qty.sort_values(ascending=False)
top_five_clients = sorted_clients.head(5)
for client_id in top_five_clients.index:
    total_spent = df[df["client_id"] == client_id]["qty"].sum()
    print("Client ID {}: Total spent by quantity: {}".format(client_id, total_spent))

Client ID 78965: Total spent by quantity: 4033302
Client ID 86721: Total spent by quantity: 1010426
Client ID 30143: Total spent by quantity: 846069
Client ID 85931: Total spent by quantity: 644423
Client ID 77354: Total spent by quantity: 506382


In [108]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
client_totals = df.groupby("client_id").agg(
    total_units_purchased=("qty", "sum"),
    total_shipping_price=("shipping_price", "sum"),
    total_revenue=("total_price", "sum"),
    total_profit=("profit", "sum")
)
top_five_clients = client_totals.nlargest(5, "total_revenue")
print(top_five_clients)

           total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
78965                    4033302                7383.3   6.500970e+08   
85931                     644423                2763.0   2.263611e+08   
38473                     402770                7832.2   1.787640e+08   
86721                    1010426                6154.6   1.109862e+08   
84344                     466968                6944.0   1.014701e+08   

           total_profit  
client_id                
78965      2.468595e+08  
85931      1.083580e+08  
38473      4.294628e+07  
86721      3.702421e+07  
84344      2.727328e+07  


In [ ]:
# Format the data and rename the columns to names suitable for presentation.
top_five_clients = top_five_clients.rename(columns={
    "total_units_purchased": "Total Units Purchased",
    "total_shipping_price": "Total Shipping Price",
    "total_revenue": "Total Revenue",
    "total_profit": "Total Profit"
    })
# Define the money columns. 
money_columns = ["Total Shipping Price", "Total Revenue", "Total Profit"]
df[money_columns] = df[money_columns].astype(float)

# Define a function that converts a dollar amount to millions.
def format_currency(value):
    return '${:,.2f}'.format(value)
for column in money_columns:
    df[column] = df[column].apply(format_currency)

# Apply the currency_format_millions function to only the money columns. 
top_five_clients["total_units_purchased"] = top_five_clients["total_units_purchased"].apply(lambda x: "{:,.0f}".format(x))
top_five_clients["total_shipping_price"] = top_five_clients["total_shipping_price"].apply(format_currency)
top_five_clients["total_revenue"] = top_five_clients["total_revenue"].apply(format_currency)
top_five_clients["total_profit"] = top_five_clients["total_profit"].apply(format_currency)

# Rename the columns to reflect the change in the money format. 
new_column_names = {
    "Total Shipping Price": "Total Shipping Price (USD)",
    "Total Revenue": "Total Revenue (USD)",
    "Total Profit": "Total Profit (USD)"
}

df = df.rename(columns=new_column_names)

print(df)

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_data = .sort_values(by="Total Profit (millions)", ascending=False)